core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -63,14 +65,14 @@ public static <T> CloseableIterable<T> filter(CloseableIterable<T> rows, Functio
     return equalityFilter.filter(rows);
   }
 
-  public static <T> CloseableIterable<T> filter(CloseableIterable<T> rows, Function<T, Long> rowToPosition,
-                                                PositionDeleteIndex deleteSet) {
-    if (deleteSet.isEmpty()) {
-      return rows;
-    }
-
-    PositionSetDeleteFilter<T> filter = new PositionSetDeleteFilter<>(rowToPosition, deleteSet);
-    return filter.filter(rows);
+  public static <T> CloseableIterable<T> markDeleted(CloseableIterable<T> rows, Predicate<T> isDeleted,
+                                                     Consumer<T> deleteMarker) {
+    return CloseableIterable.transform(rows, row -> {
+      if (isDeleted.test(row)) {
```

Comment:
My point is we have to check each row inevitably for non-vectorized read. The vectorized read is a different story. For example, we can generate the isDelete boolean array directly from the pos deletes. 
Of course there is no way to avoid per-row check for eq deletes, neither with vectorized read, nor non-vectorized read.

Reply:
Is the worry here that branch prediction is gonna be bad here? I would assume that speculative branch execution would probably make ay optimizations we do here very sensitive to the amount of deletes. Anyway I'm on team leave it simple for now.